KAFKA-19589: Check for ability to skip position validation in application thread when collecting buffered data

kirktrue opened the pull request:
Introduces `PositionsValidator` which queries the assignment data  from
the `SubscriptionState` from the  application thread, allowing
`AsyncKafkaConsumer` to potentially avoid the need to wait.

The impetus for this change is the observation of these two points:

1. Waiting for the completion of
`OffsetsRequestManager.updateFetchPositions()` on the application thread
means either busy-waiting or blocking, either of which add significant
(~60%) CPU load to the `AsyncKafkaConsumer` compared to the
`ClassicKafkaConsumer`
2. In testing, data shows that 99.99+% of the time that
`OffsetsRequestManager.updateFetchPositions()` is called, the partitions
are up-to-date and there is no need to fetch offsets.

This patch allows the check for stable partitions to be made in the
application thread, resulting in far less waiting in the critical path
of `AsyncKafkaConsumer.poll()`.

Reviewers: Lianet Magrans <lmagrans@confluent.io>, Andrew Schofield
 <aschofield@confluent.io>

kirktrue committed:
WIP: Attempting to short-circuit event call in updateFetchPositions()

kirktrue committed:
Experiment: skip the call to updateFetchPositions by accessing shared, thread-safe state

kirktrue committed:
Setting cachedSubscriptionHasAllFetchPositions to false when assignments change

kirktrue committed:
Reverting use of sharedOffsetsState.offsetFetcherUtils() and retrieving the OffsetFetcherUtils once, in the constructor

kirktrue committed:
Renaming to CommitOffsetsSharedState

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue commented:
# Results for August 9, 2025 testing

## Metrics

The metrics shown below come from three separate test runs:

1. Test run `CLASSIC` uses the `CLASSIC` group protocol from `trunk`
2. Test run `CONSUMER` uses the `CONSUMER` group protocol, also from `trunk`
3. Test run `CONSUMER (branch)` uses the `CONSUMER` group protocol from my current development branch (`KAFKA-19589-reduce-events-in-update-fetch-positions`)

Results:

| Metric                    |                   CLASSIC |          CONSUMER (trunk) |         CONSUMER (branch) |
|:--------------------------|--------------------------:|--------------------------:|--------------------------:|
| bytes-consumed-rate       |             162857862.759 |              48074181.128 |              77237286.006 |
| bytes-consumed-total      |          104199838490.000 |          104199838490.000 |          104199838490.000 |
| fetch-latency-avg         |                    27.563 |                    47.571 |                    25.268 |
| fetch-latency-max         |                   502.000 |                   507.000 |                   506.000 |
| fetch-rate                |                   161.195 |                    51.055 |                    78.003 |
| fetch-size-avg            |               1023854.536 |               1023506.132 |               1023205.444 |
| fetch-size-max            |               1033664.000 |               1033664.000 |               1033664.000 |
| fetch-throttle-time-avg   |                     0.000 |                     0.000 |                     0.000 |
| fetch-throttle-time-max   |                     0.000 |                     0.000 |                     0.000 |
| fetch-total               |                101908.000 |                102036.000 |                101959.000 |
| records-consumed-rate     |                312587.069 |                 92272.900 |                148248.150 |
| records-consumed-total    |             199999690.000 |             199999690.000 |             199999690.000 |
| records-lag-max           |               4484207.000 |               1806989.000 |               2578731.000 |
| records-lead-min          |              29035357.000 |              32266785.000 |              31534419.000 |
| records-per-request-avg   |                  1965.172 |                  1964.503 |                  1963.926 |
| commit-sync-time-ns-total |                     0.000 |                     0.000 |                     0.000 |
| committed-time-ns-total   |                     0.000 |                     0.000 |                     0.000 |
| incoming-byte-rate        |             163876206.099 |              48227380.969 |              77529673.969 |
| incoming-byte-total       |          104604394434.000 |          104604441109.000 |          104622899962.000 |
| io-ratio                  |                     0.244 |                     0.092 |                     0.146 |
| io-time-ns-avg            |                  3865.850 |                  2363.981 |                  3871.983 |
| io-time-ns-total          |          157231567818.000 |          217749720755.000 |          195497217580.000 |
| io-wait-ratio             |                     0.303 |                     0.523 |                     0.503 |
| io-wait-time-ns-avg       |                  4784.395 |                 13368.476 |                 13311.075 |
| io-wait-time-ns-total     |          111080698186.000 |         1148273922014.000 |          655479798988.000 |
| last-poll-seconds-ago     |                     0.000 |                     0.000 |                     0.000 |
| network-io-rate           |                  2213.585 |                   689.148 |                  1100.397 |
| network-io-total          |               1385915.000 |               1426966.000 |               1522117.000 |
| outgoing-byte-rate        |                 17912.864 |                  5518.740 |                  8596.430 |
| outgoing-byte-total       |              11354966.000 |              11442918.000 |              11387680.000 |
| poll-idle-ratio-avg       |                       NaN |                       NaN |                       NaN |
| request-rate              |                   162.060 |                    51.471 |                    78.488 |
| request-size-avg          |                   110.532 |                   107.220 |                   109.526 |
| request-size-max          |                   238.000 |                   203.000 |                   203.000 |
| request-total             |                102231.000 |                102940.000 |                102512.000 |
| response-rate             |                   162.013 |                    51.441 |                    78.424 |
| response-total            |                102226.000 |                102935.000 |                102507.000 |
| select-rate               |                 63238.109 |                 39094.742 |                 37796.679 |
| select-total              |              40007185.000 |              87649140.000 |              50222746.000 |
| time-between-poll-avg     |                     0.016 |                     0.054 |                     0.034 |
| time-between-poll-max     |                   100.000 |                   100.000 |                    84.000 |
| Average CPU load          |                    92.23% |                   146.39% |                   149.89% |
| P99 CPU load              |                    93.50% |                   148.00% |                   151.00% |
| Max CPU load              |                    93.50% |                   148.00% |                   151.00% |

## Execution

The test is run on a single AWS instance targeting a six-node cluster. The data on the topic is pre-populated with 200,000,000 messages of 512 bytes each before any of the tests are run. A single “warm up” run of the performance test is executed before executing the three test runs mentioned above.

The command to execute the test is:

```bash
kafka-consumer-perf-test.sh \
  --bootstrap-server $BOOTSTRAP_SERVER \
  --topic $TOPIC_NAME \
  --messages $NUM_MESSAGES \
  --consumer.config conf/consumer-$GROUP_PROTOCOL.properties \
  --show-detailed-stats \
  --print-metrics
```

`NUM_MESSAGES` is set to `200000000` (200 million).

## Client configuration

The consumer configuration sets `max.poll.records=5`, but is otherwise mostly defaults apart from connectivity/authentication:

```
bootstrap.servers=$BOOTSTRAP_SERVER
sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required
  username="$USER_ID"
  password='$USER_PASSWORD' ;
security.protocol=SASL_SSL
sasl.mechanism=PLAIN
ssl.endpoint.identification.algorithm=
client.dns.lookup=use_all_dns_ips

# Key
max.poll.records=5
```

## Files

[cpu-classic-trunk.txt](https://github.com/user-attachments/files/21703479/cpu-classic-trunk.txt)
[cpu-consumer-branch.txt](https://github.com/user-attachments/files/21703480/cpu-consumer-branch.txt)
[cpu-consumer-trunk.txt](https://github.com/user-attachments/files/21703481/cpu-consumer-trunk.txt)

[perf-classic-trunk.txt](https://github.com/user-attachments/files/21703482/perf-classic-trunk.txt)
[perf-consumer-branch.txt](https://github.com/user-attachments/files/21703483/perf-consumer-branch.txt)
[perf-consumer-trunk.txt](https://github.com/user-attachments/files/21703484/perf-consumer-trunk.txt)

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java:
```diff
@@ -243,8 +237,11 @@ public CompletableFuture<Boolean> updateFetchPositions(long deadlineMs) {
 
             if (subscriptionState.hasAllFetchPositions()) {
                 // All positions are already available
+                commitOffsetsSharedState.setSubscriptionHasAllFetchPositions(true);
```
uhm updating the cached "subscriptionHasAllPositions" here will run before fetching. Wouldn't we have an outdated cache if the positions become invalid after the fetch result? (ex. fetch out of range that will reset offsets). 

In such cases, I imagine the positions would become invalid after this point, when processing the fetch result, but the cache will remain with hasAllPositions true and wrongly skip `updateFetchPositions` on the next poll. Am I missing something? I see we only update the cache here and when the assignment changes.

frankvicky replied:
> What about continuing relying on the source of truth (subscriptionState.hasAllFetchPositions), but piggyback on the PollEvent to get it safely (we need that event anyways).

Agree. The new class has introduced additional complexity; it would be nice if we could leverage the current event.
IIUC, `CommitOffsetsSharedState` appears to act as a signal between the app thread and the network thread. If so, it will serve the same purpose as the `ApplicationEvent`.

kirktrue replied:
@frankvicky—Agreed. This is complexity that I wish we didn't need to introduce. Unfortunately, coordinating events between the application and background thread are adding substantial CPU load overhead. The process of scheduling the synchronization is very much a bottleneck.

However, `CommitOffsetsSharedState` is _not_ an event specifically for the purpose that we can then avoid signaling, synchronization, or any other inter-thread communication.

lianetm reviewed (commented):
Hey, thanks for the PR!

Agree on the intention of reducing the hops between the threads, but high level comment on the approach for this case. Introducing this cache to know about having all positions is tricky because we need to maintain it from all places where assignment or position state changes (see comment). 

What about continuing relying on the source of truth (`subscriptionState.hasAllFetchPositions`), but piggyback on the `PollEvent` to get it safely (we need that event anyways). 

We already piggyback on it for something similar, to check we start fetching after having triggered the auto-commit with the current positions (`pollEvent.reconcileAndAutoCommit`).  Would something similar work to check we have all positions, reusing that same hop/event?

kirktrue committed:
Update to check the SubscriptionState instead of relying solely on the cached value

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitOffsetsSharedState.java:
```diff
@@ -0,0 +1,162 @@
+ * sparingly in both the application and background threads. Both of those classes are heavily synchronized given
+ * their use by the {@link ClassicKafkaConsumer}, so their use in a multithreaded fashion is already established.
+ */
+public class CommitOffsetsSharedState implements MemberStateListener {
```
This component seems to be for `OffsetsState` (could be committed offsets or partition offsets), correct? If so, should we revise the name?

kirktrue replied:
Generalized the name to `SharedConsumerState` for reuse elsewhere.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitOffsetsSharedState.java:
```diff
@@ -0,0 +1,162 @@
+
+        // If the cached value is set and there are no partitions in the AWAIT_RESET, AWAIT_VALIDATION, or
+        // INITIALIZING states, it's ok to skip.
+        if (cachedSubscriptionHasAllFetchPositions.get() && offsetFetcherUtils.getPartitionsToValidate().isEmpty() && subscriptions.hasAllFetchPositions())
```
Why relying on the extra cache here (`cachedSubscriptionHasAllFetchPositions.get()`)? Wouldn't this mean that we could potentially reuse the cached value for hasAllFetchPositions across multiple poll iterations? 

We only want to reuse it within the same poll iteration I expect (which is what both consumers do before this PR: within a single poll iteration, always check `subscriptions.hasAllFetchPositions` the first time it's needed, and then reuse the value to avoid checking it more than once in that same poll)  

Also, I would expect this check is redundant: `offsetFetcherUtils.getPartitionsToValidate().isEmpty() && subscriptions.hasAllFetchPositions()` 
(if there are partitions to validate, hasAllFetchPositions will be false, because the `AWAIT_VALIDATION` state `hasValidPosition` is false

kirktrue replied:
Removed use of `cachedSubscriptionHasAllFetchPositions`. There are probably some savings to be had by keeping it and fixing up the issues, but they're much more minor than the event generation.

kirktrue committed:
Remove unused state listener logic from CommitOffsetsSharedState

Eliminates the MemberStateListener implementation and related state caching from CommitOffsetsSharedState, simplifying its logic. Updates OffsetsRequestManager, RequestManagers, and associated tests to reflect the removal of this listener and its state management.

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue committed:
Refactored so that NetworkClientDelegate's metadata error was available to the shared consumer state object

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue committed:
Clearing the metadata error exception when thrown

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue committed:
Added comments and reverted unnecessary changes

kirktrue committed:
Ugh. Typo

kirktrue commented:
@chia7712 @frankvicky—would you or someone you know to be familiar with the new `AsyncKafkaConsumer` be willing to help review this PR? Thanks!

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue commented:
FYI: test failures appear related to the broader flakiness of the tests over the last seven days:

- `AclControlManagerTest.testDeleteExceedsMaxRecords()`: [10% flaky](https://develocity.apache.org/scans/tests?search.rootProjectNames=kafka&search.timeZoneId=America%2FLos_Angeles&tests.container=org.apache.kafka.controller.AclControlManagerTest&tests.sortField=FLAKY)
- `DescribeStreamsGroupTest.testDescribeStreamsGroupWithShortTimeout()`: [26% flaky](https://develocity.apache.org/scans/tests?search.rootProjectNames=kafka&search.timeZoneId=America%2FLos_Angeles&tests.container=org.apache.kafka.tools.streams.DescribeStreamsGroupTest&tests.sortField=FLAKY)

AndrewJSchofield commented:
> FYI: test failures appear related to the broader flakiness of the tests over the last seven days:
> 
> * `AclControlManagerTest.testDeleteExceedsMaxRecords()`: [10% flaky](https://develocity.apache.org/scans/tests?search.rootProjectNames=kafka&search.timeZoneId=America%2FLos_Angeles&tests.container=org.apache.kafka.controller.AclControlManagerTest&tests.sortField=FLAKY)
> * `DescribeStreamsGroupTest.testDescribeStreamsGroupWithShortTimeout()`: [26% flaky](https://develocity.apache.org/scans/tests?search.rootProjectNames=kafka&search.timeZoneId=America%2FLos_Angeles&tests.container=org.apache.kafka.tools.streams.DescribeStreamsGroupTest&tests.sortField=FLAKY)

Please put in an issue and PR that marks these two tests as `@Flaky`.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SharedConsumerState.java:
```diff
@@ -0,0 +1,151 @@
+
+/**
+ * This class stores shared state needed by both the application thread ({@link AsyncKafkaConsumer}) and the
+ * network thread ({@link ConsumerNetworkThread}) to avoid costly inter-thread communication, where possible.
```
nit: We have used "background thread" rather than "network thread" everywhere else. And I know the class name is different, but I would go for consistency in the comments so people do not get confused about a new kind of thread.

kirktrue replied:
Out of curiosity I looked up the number of instances of "background thread" vs. "network thread" in the client source:

```
$ grep -R "network thread" clients/src/main/java | wc -l
      61
$ grep -R "background thread" clients/src/main/java | wc -l
      42
```

That's not exact, of course, but there are clearly two warring factions fighting for dominance here 😆

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SharedConsumerState.java:
```diff
@@ -0,0 +1,151 @@
+     * method will return {@code false}. Otherwise, this method will return {@code true}, which signals to the
+     * application thread that the {@link CheckAndUpdatePositionsEvent} can be skipped.
+     *
+     * @return true if all checks pass, false if either of the latter two checks fail
```
I think an important part of this method's behaviour is that it will throw exceptions in several situations. That's probably relevant in the javadoc comment on this method.

kirktrue replied:
I've added verbiage related to the exceptions to the `canSkipUpdateFetchPositions()` method JavaDoc.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SharedConsumerState.java:
```diff
@@ -0,0 +1,151 @@
+        return offsetFetcherUtils;
+    }
+
+    public SharedExceptionReference updatePositionsError() {
```
So, this sounds a bit odd because it sounds like a mutating method name. Does it update the positions error? Nope. Would positionsUpdateError be acceptable?

kirktrue replied:
You're right. I've stared at `updatePositionsError` for so long and hard that its name being confusing didn't jump out at me. I've changed it to `positionsUpdateError` here and where it's referenced by the callers.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -1847,11 +1871,21 @@ private Fetch<K, V> collectFetch() {
             CheckAndUpdatePositionsEvent checkAndUpdatePositionsEvent = new CheckAndUpdatePositionsEvent(calculateDeadlineMs(timer));
             wakeupTrigger.setActiveTask(checkAndUpdatePositionsEvent.future());
-            cachedSubscriptionHasAllFetchPositions = applicationEventHandler.addAndGet(checkAndUpdatePositionsEvent);
+            applicationEventHandler.addAndGet(checkAndUpdatePositionsEvent);
```
We are discarding the result of this event. Does it need a result any more? If not, do we want the result?

kirktrue replied:
Good point. I've removed the returned value from `CheckAndUpdatePositionsEvent` because we're no longer caching the result of the `hasAllFetchPositions()` anyway.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. Overall, I think my main observation about the PR is that it makes the class diagram even more peculiar than it was before. The classes know too much about each other really. Whether that's a future refactor or a refactor as part of this PR I am not sure.

kirktrue committed:
Refactoring and JavaDoc clarification

kirktrue committed:
Removed superfluous boolean return value from CheckAndUpdatePositionsEvent

kirktrue committed:
Updated CheckAndUpdatePositionsEvent doc to specify that event completion criteria

kirktrue commented:
Thanks for the review @AndrewJSchofield!

> Overall, I think my main observation about the PR is that it makes the class diagram even more peculiar than it was before. The classes know too much about each other really. Whether that's a future refactor or a refactor as part of this PR I am not sure.

I wholeheartedly agree. Do you think that grouping all the shared state under `ThreadSafeConsumerState` (and its subclasses) is a good idea or bad? I'm struggling with a) keeping the pieces loosely-coupled, and b) keeping the pieces together under an umbrella where the intention/hazards are clear.

Any thoughts on this are welcome!

Thanks!

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ValidatePositionsChecker.java:
```diff
@@ -0,0 +1,160 @@
+        // In cases of metadata updates, getPartitionsToValidate() will review the partitions and
+        // determine which, if any, need to be validated. If any partitions require validation, the
+        // update fetch positions step can't be skipped.
+        if (!offsetFetcherUtils.getPartitionsToValidate().isEmpty())
```
This means we would move positions validation to the app thread, which seems tricky.  

I expect we could have a race condition here if a partition awaiting validation is revoked (in the background) while this runs (in the app thread now), mainly because this get the `partitionsNeedingValidation` and then it's positions in separate steps, so I guess it could fail with "No current assignment for partition x" when trying to get `subscriptionState.position(tp)` (here https://github.com/apache/kafka/blob/2ba30cc466fd854b6ed7fa2afd157a80fcf7382a/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java#L182-L184)

kirktrue replied:
Taking step back from the current implementation for a moment, what I want is this API:

```java
public interface EventExecutionChecker {

    public boolean shouldExecuteEvent();

}
```

Implementations of this interface have the following requirements:

* **Must** be thread-safe, for use by the application thread
* **Must not** return false negatives
* **May** return false positives; "accidentally" executing the event even when not strictly necessary is OK
* **Must not** mutate state

Intuitively, this seems possible. That's what I'd like to explore.

kirktrue replied:
I changed `subscriptionState.position(tp)` to `subscriptionState.positionOrNull(tp)`, in addition.

lianetm replied:
I think we still need to be more careful with this call. 

We just need to check if there are partition that need validation, but we're doing much more than that, mutating the subscription state. The issue to me is truly the `validatePositionsOnMetadataChange` called from within  `offsetFetcherUtils.getPartitionsToValidate()`. `validatePositionsOnMetadataChange` can change the partition state and even the position. We don't want to be doing that from the app thread.

So what about having a lightweight version of getPartitionsToValidate to really just check in the app thread, and leaving mutation to the background (keeping the original one, it's used elsewhere) 
- check the cachedValidatePositionsException (atomic, no concerns clearing it too)
- check if the metadata version changed (instead of the full validatePositionsOnMetadataChange, if it changed we cannot skip in the app thread and should go to the background to fully validatePositionsOnMetadataChange)
- check the subscriptionState.partitionsNeedingValidation

Wdyt? I'm still going over this

kirktrue replied:
I've refactored the code so that no mutations are made on the application thread.

lianetm reviewed (commented):
Thanks for the updates! I'm afraid we are still introducing gaps (comment below) by splitting the logic of the existing `CheckAndUpdatePositionsEvent` event into the app thread and background thread, and honestly I don't see how we could sort them out cleanly and confidently with this split approach. But let's not surrender :) 

Actually not only for this PR but also the one attempting a similar approach for the `PollEvent`, I wonder if instead of splitting current events logic into the app and background thread, we should explore consolidating them in the background, closer to the fetch logic.  This is my reasoning:

If we take a step back, both `CheckAndUpdatePositionsEvent` and `PollEvent` do stuff that only matters for fetching (we use them in the app thread, in a blocking way, only to ensure things happen before the actual fetching, which in the end runs in the background too).

This is what we currently have (before this PR):
- `PollEvent` to trigger reconciliation and auto-commit -> must happen before updating positions
- `CheckAndUpdatePositionsEvent` to throw previous errors and update positions -> must happen before fetching
- `CreateFetchRequestsEvent` -> triggers the actual fetching

So what about we move it all to the fetch in the background, and when processing a `CreateFetchRequestEvent` we:
1. trigger reconciliation & auto-commit (what the PollEvent used to do)
2. update fetch positions (what the update fetch positions used to do)
3. trigger fetch 

Note we still need to throw previous exceptions related to the `updateFetchPositions` before returning data from the buffer, but that part can be safely done in the app thread using AtomicRef<Throwable> at the beginning of the poll. There could be other things to consider, just high level ideas for now in case it helps. Thoughts?

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue commented:
> Thanks for the updates! I'm afraid we are still introducing gaps (comment below) by splitting the logic of the existing `CheckAndUpdatePositionsEvent` event into the app thread and background thread, and honestly I don't see how we could sort them out cleanly and confidently with this split approach. But let's not surrender :)
> 
> Actually not only for this PR but also the one attempting a similar approach for the `PollEvent`, I wonder if instead of splitting current events logic into the app and background thread, we should explore consolidating them in the background, closer to the fetch logic. This is my reasoning:
> 
> If we take a step back, both `CheckAndUpdatePositionsEvent` and `PollEvent` do stuff that only matters for fetching (we use them in the app thread, in a blocking way, only to ensure things happen before the actual fetching, which in the end runs in the background too).
> 
> This is what we currently have (before this PR):
> 
> * `PollEvent` to trigger reconciliation and auto-commit -> must happen before updating positions
> * `CheckAndUpdatePositionsEvent` to throw previous errors and update positions -> must happen before fetching
> * `CreateFetchRequestsEvent` -> triggers the actual fetching
> 
> So what about we move it all to the fetch in the background, and when processing a `CreateFetchRequestEvent` we:
> 
> 1. trigger reconciliation & auto-commit (what the PollEvent used to do)
> 2. update fetch positions (what the update fetch positions used to do)
> 3. trigger fetch
> 
> Note we still need to throw previous exceptions related to the `updateFetchPositions` before returning data from the buffer, but that part can be safely done in the app thread using AtomicRef at the beginning of the poll. There could be other things to consider, just high level ideas for now in case it helps. Thoughts?

Thanks for the explanation, @lianetm!

I definitely like the idea of having a consolidated `DoEverythingNeededForFetchEvent` that handles the logic that's currently split up across `PollEvent`, `CheckAndUpdatePositionsEvent`, and `CreateFetchRequestsEvent`. Unfortunately, I see two constraints that prevent the ability to consolidate those events:

1. Offset commit callbacks, background events, and collecting fetch data must be executed in the application thread
2. The execution of those callbacks and events in the application thread are interleaved with the event logic executed in the background thread for `PollEvent`, `CheckAndUpdatePositionsEvent`, and `CreateFetchRequestsEvent`

If there's a way around these two issues, there's hope. I just don't yet see a way around those constraints without a major change 😞

kirktrue commented:
> I definitely like the idea of having a consolidated `DoEverythingNeededForFetchEvent` that handles the logic that's currently split up across `PollEvent`, `CheckAndUpdatePositionsEvent`, and `CreateFetchRequestsEvent`. Unfortunately, I see two constraints that prevent the ability to consolidate those events:
> 
> 1. Offset commit callbacks, background events, and collecting fetch data must be executed in the application thread
> 2. The execution of those callbacks and events in the application thread are interleaved with the event logic executed in the background thread for `PollEvent`, `CheckAndUpdatePositionsEvent`, and `CreateFetchRequestsEvent`
> 
> If there's a way around these two issues, there's hope. I just don't yet see a way around those constraints without a major change 😞

In thinking about the consolidated event approach a little more, I think there's a possible way forward. Here are some assumptions:

1. I _assume_ that the majority of `poll()` executions execute neither async commit offset callbacks nor background events. i.e., I _assume_ these are _comparatively_ rare when it comes to polling.
2. I _assume_ that there's a way to discern, from the background thread, if we need to execute async commit offset callbacks nor background events.
3. I _assume_ the `DoEverythingNeededForFetchEvent` event processing could be prematurely exit if it detects the need for the application thread’s help.

Given the above, can we introduce `DoEverythingNeededForFetchEvent` as an optimized "fast path" that can execute the bulk of fetch uninterrupted in the background thread, but still have a "slow path" for when we need to fall back to using one or both of `PollEvent` or `CheckAndUpdatePositionsEvent`, when needed?

There's still another piece to the puzzle that I haven't figured out yet, which is how to handle the `CreateFetchRequestsEvent`. The `FetchCollector.collectFetch()` logic runs on the application thread because it may need to decrypt and/or decompress data from the buffer to determine if there are any returnable records. Maybe we just assume that if there's any data in the fetch buffer that we shouldn't execute the `CreateFetchRequestsEvent` logic? Can we _assume_ that most of the time if there's data in the fetch buffer that it's usable, and if we later try to get it out and it's bogus, we just have to fall back to the slow path?

lianetm commented:
> Given the above, can we introduce DoEverythingNeededForFetchEvent as an optimized "fast path" that can execute the bulk of fetch uninterrupted in the background thread, but still have a "slow path" for when we need to fall back to using one or both of PollEvent or CheckAndUpdatePositionsEvent, when needed?

Well, I was hoping we could move to a simplified fast path only, consolidating the fetch logic in the background. Single event to trigger fetch (including all the pre-fetch actions, and triggering fetch), and keeping callbacks triggering in the app thread. This is what I had in mind to explore:

- Poll event -> remove and consolidate within `CreateFetchRequestEvent` (to trigger reconciliation and auto-commit before updating positions to fetch)
- updateAssignmentMetadataIfNeeded -> extract what needs to happen in the app thread and leave it there. Consolidate fetch-related logic within `CreateFetchRequestEvent`
    - Commit callbacks: remain in app thread, triggered via `offsetCommitCallbackInvoker`
    - `UpdatePatternSubscriptionEvent`: this logic simply needs to run before joining a group, and we already have a hook for that in the background (`membershipMgr.onConsumerPoll`), so we could consider removing this event and move the logic there.
    - Rebalance callbacks: remain in app thread, triggered via `processBackgroundEvents` (this is unrelated to the previous `UpdatePatternSubscriptionEvent` logic so no concerns with the order here I would say) 
    - updateFetchPositions: consolidate within `CreateFetchRequestEvent`

Would that work? I'm still thinking about it myself, but if it works, it would be a nice simplification to the current poll(), consolidating fetch in the background with a single `DoEverythingNeededForFetchEvent`/`TriggerFetch` event

Then, regarding:
> how to handle the CreateFetchRequestsEvent. The FetchCollector.collectFetch() logic runs on the application thread

I don't quite get the concern, probably missing something? We are just changing how we prep before sending a fetch (consolidating in a single event), but no changes on how we wait for the `CreateFetchRequestsEvent` (we will continue to wait for it), and no changes on how we take data out of the buffer (collectFetch, in the app thread, etc, agreed).

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue commented:
I'd like to realign the core conversation to the approach in this PR (and #20363) which is the elimination of `addAndGet()` on the common path. The performance bottleneck in `AsyncKafkaConsumer.poll()` is `addAndGet()`, pure and simple. When the application thread calls `addAndGet()` on the common path in `poll()`, the CPU usage spikes. I'm convinced that we can remove calls to `addAndGet()` from the common path in `poll()` through smart and careful sharing of state.

AndrewJSchofield commented:
> I'd like to realign the core conversation to the approach in this PR (and #20363) which is the elimination of `addAndGet()` on the common path. The performance bottleneck in `AsyncKafkaConsumer.poll()` is `addAndGet()`, pure and simple. When the application thread calls `addAndGet()` on the common path in `poll()`, the CPU usage spikes. I'm convinced that we can remove calls to `addAndGet()` from the common path in `poll()` through smart and careful sharing of state.

I agree with this, but I should point of that it's going to be quite a big shift. In the share consumer, we do have an event to begin fetching, but the background thread just continues to fetch until it receives data. It doesn't need to be repeatedly prodded. We also do not have events for messing around with the position. The background thread could be a bit more autonomous with less requirement to give it synchronous pieces of work to do.

kirktrue commented:
I'm closing this PR as we have decided against going any further with this approach due to thread safety concerns.

cc @lianetm @AndrewJSchofield

kirktrue commented:
I'm closing this PR as we have decided against going any further with this approach due to thread safety concerns.

cc @lianetm @AndrewJSchofield

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue committed:
Removing references to shared errors

kirktrue committed:
More changes to reduce diffs in PR

kirktrue committed:
Update NetworkClientDelegate.java

kirktrue committed:
Update AsyncKafkaConsumerTest.java

kirktrue committed:
Renaming ThreadSafeAsyncConsumerState to ValidatePositionsChecker

kirktrue committed:
Removed unnecessary imports

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue committed:
Updated partitionsNeedingValidation to perform all filtering logic in one go and updated relevant tests

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

kirktrue committed:
Refactored to avoid mutations on the application thread

kirktrue committed:
Minor whitespace change

kirktrue committed:
Reverted unnecessary change

kirktrue committed:
Minor refactoring

kirktrue committed:
Revert now-unnecessary change in AsyncKafkaConsumerTest

kirktrue committed:
Removed unused variable

kirktrue committed:
Added a linebreak in import list to make linter happy

kirktrue committed:
Merge branch 'trunk' into KAFKA-19589-reduce-events-in-update-fetch-positions

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PositionsValidator.java:
```diff
@@ -0,0 +1,157 @@
+     * This method is called by the background thread in response to {@link AsyncPollEvent} and
+     * {@link CheckAndUpdatePositionsEvent}.
+     */
+    Map<TopicPartition, SubscriptionState.FetchPosition> getPartitionsToValidate(ApiVersions apiVersions) {
```
should we rename this to a kind of refreshAndGetPartitionsToValidate? it's truly hidden here that this func does update the partition state and even maybe the position

kirktrue replied:
Renamed.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PositionsValidator.java:
```diff
@@ -0,0 +1,157 @@
+        // In cases of metadata updates, getPartitionsToValidate() will review the partitions and
+        // determine which, if any, need to be validated. If any partitions require validation, the
+        // update fetch positions step can't be skipped.
+        if (subscriptions.hasPartitionsNeedingValidation(time.milliseconds())) {
```
do we really need this or could we remove and rely solely on the `hasAllFetchPositions`? (both checks iterate the full assignment and run on each poll, so better to go with the minimum) 

If there are partitions needing validation, the `hasAllFetchPositions` would  return false anyways (because the position state would be AWAITING_VALIDATION). Would it be enough?

Also, the `hasPartitionsNeedingValidation` considers backoff, but that is not something we want to consider when checking to skip or not here (if the partition is AWAITING_VALIDATION, we can't skip, even if it's awaiting backoff). We need to consider backoff only when actually sending the validation request in the background

kirktrue replied:
I'm not sure the check for the metadata version as it is now is sufficient. The application thread and background thread both check the metadata version. Isn't it possible for the application thread version check to be made right before the metadata is updated on the background thread?

lianetm replied:
oh totally agree that the metadata check alone is not enough, I meant that if the metadata +  `subscriptions.hasAllFetchPositions()` check is enough and we can get rid of the `hasPartitionsNeedingValidation` ? 

Seemed to me that what we get from `hasPartitionsNeedingValidation` (break if there are partitions AWAITING_VALIDATION), is included in the  `hasAllFetchPositions` (if the partition is AWAITING_VALIDATION, hasValidPosition is false).

kirktrue replied:
> Seemed to me that what we get from `hasPartitionsNeedingValidation` (break if there are partitions AWAITING_VALIDATION), is included in the `hasAllFetchPositions` (if the partition is AWAITING_VALIDATION, hasValidPosition is false).

Agreed.

kirktrue replied:
Regarding the metadata version check, I'm concerned that it's insufficient to protect against the metadata changing immediately after `canSkipUpdateFetchPositions()` returns `true`:

```java
boolean canSkipUpdateFetchPositions() {
    maybeThrowError();

    if (metadataUpdateVersion.get() != metadata.updateVersion()) {
        return false;
    }

    // ----------------------------------------------------------------
    // The background thread could update the metadata here
    // ----------------------------------------------------------------

    return subscriptions.hasAllFetchPositions();
}
```

I don't think the check stands up to this sequence of events:

1. Application thread: `Consumer.poll()` is invoked
2. Application thread: a new `AsyncPollEvent` is enqueued
3. Application thread: the metadata version number check above is made
4. Background thread: `NetworkClient.handleCompletedReceives()` is called (after the socket is awoken in step 2), detects the metadata response, and updates the metadata, bumping the version number
5. Application thread: pulls data out of the `FetchBuffer` and returns to the user
7. Background thread: `ApplicationEventProcessor.process(AsycPollEvent)` calls   `validatePositionsOnMetadataChange()` detects the metadata version change and marks some of the positions as needing validation

lianetm replied:
yes, agree, but that situation happens no matter what checks we do here (ex. would happen also with the previous check of `hasPartitionsNeedingValidation`), and was possible also before this PR actually, when we had a blocking poll with `CheckAndUpdatePositions` (metadata could change right after the app thread regained control after CheckAndUpdatePositions, but the app thread would still collectFetch, right?)

The good thing is that the `collectFetch` does check the basics for correctness before returning data from the buffer (partition still assigned, with valid position). So if the metadata did change in the background and the position was marked as invalid the data won't be returned, if not, I expect it's safe to return (even though there may be a metadata change in-progress let's say).

kirktrue replied:
> The good thing is that the collectFetch does check the basics for correctness before returning data from the buffer (partition still assigned, with valid position). So if the metadata did change in the background and the position was marked as invalid the data won't be returned, if not, I expect it's safe to return (even though there may be a metadata change in-progress let's say).

The concern is when the application thread completes execution of `collectFetch()` before the background thread has even starts executing `validatePositions()`.

I have to look again at what happens to data in the fetch buffer for a partition that is marked as 'needing validation.'

lianetm commented:
Another gap I noticed, not when collecting from the buffer, but when sending pre-fetches (so it's not directly with the changes of this PR, but related). 

Here we're ensuring we collect data from the buffer only when there is no need to validate positions (or when positions were validated).
Following a similar principle, I expect we should ensure pre-fetches are only built when we have updated the partition state (metadata changed? needs validation?). The classic consumer ensures that here :
https://github.com/apache/kafka/blob/dc9ae8bf805a93896104a51417318729e032fc3e/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ClassicKafkaConsumer.java#L674-L675
 
But I don't see how we are ensuring that on the async? we send pre-fetches by directly triggering a `CreateFetchRequestsEvent`, which will generate fetch requests without `validatePositionsOnMetadataChange`. The gap seems clear when we look at the usages of the `fetchRequestManager.createFetchRequests()`
- from AsyncPollEvent -> here it's ok, we only createFetchRequests after having updated positions
- from process(CreateFetchRequestsEvent) -> this is the gap I'm seeing, triggered for pre-fetching (could we include partitions that may need validation but since we didnt; call validatePositionsOnMetadataChange we don't notice?)

Am I missing something or is it indeed a gap?

kirktrue committed:
Renamed getPartitionsToValidate to refreshAndGetPartitionsToValidate

kirktrue committed:
Removed check for hasPartitionsNeedingValidation()

kirktrue commented:
> Another gap I noticed, not when collecting from the buffer, but when sending pre-fetches (so it's not directly with the changes of this PR, but related).
> 
> Here we're ensuring we collect data from the buffer only when there is no need to validate positions (or when positions were validated). Following a similar principle, I expect we should ensure pre-fetches are only built when we have updated the partition state (metadata changed? needs validation?). The classic consumer ensures that here :
> 
> https://github.com/apache/kafka/blob/dc9ae8bf805a93896104a51417318729e032fc3e/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ClassicKafkaConsumer.java#L674-L675
> 
> But I don't see how we are ensuring that on the async? we send pre-fetches by directly triggering a `CreateFetchRequestsEvent`, which will generate fetch requests without `validatePositionsOnMetadataChange`. The gap seems clear when we look at the usages of the `fetchRequestManager.createFetchRequests()`
> 
> * from AsyncPollEvent -> here it's ok, we only createFetchRequests after having updated positions
> * from process(CreateFetchRequestsEvent) -> this is the gap I'm seeing, triggered for pre-fetching (could we include partitions that may need validation but since we didnt; call validatePositionsOnMetadataChange we don't notice?)
> 
> Am I missing something or is it indeed a gap?

Yes, it's possible a `CreateFetchRequestsEvent` may be _enqueued_ before the processing for the `AsyncPollEvent` has completed, but the background thread doesn't start _processing_ the `CreateFetchRequestsEvent` immediately. The background processing of `AsyncPollEvent` executes without pause until after the `validatePositionsIfNeeded()` method returns. So the background thread won't _process_ the `CreateFetchRequestsEvent` until after the `AsyncPollEvent` processing (including `validatePositionsIfNeeded()`) has completed.

lianetm commented:
Thanks for the response on the pre-fetches! Positions will be marked as requiring validation before generating the request because we process app event (AsyncPoll) before polling the fetch manager, all good. 

Any concerns on your side on the last answer [here](https://github.com/apache/kafka/pull/20324#discussion_r2550854575)?

kirktrue commented:
> Any concerns on your side on the last answer [here](https://github.com/apache/kafka/pull/20324#discussion_r2550854575)?

As we'd discussed offline, we pull data from the fetch buffer on the application thread. Even without this PR, there's always a chance that something could happen in the background thread that invalidates the positions. The `FetchCollector` mitigates against those, so my concerns have been alleviated.

lianetm reviewed (approved):
Thanks @kirktrue , LGTM.

Should we update the PR title to clarify it is about checking ability to skip position validation in application thread **when collecting buffered data**"? (it's an important diff and reduced scope, we truly cannot "skip pos validation in the app thread" in general)

kirktrue commented:
> Should we update the PR title to clarify it is about checking ability to skip position validation in application thread **when collecting buffered data**"? (it's an important diff and reduced scope, we truly cannot "skip pos validation in the app thread" in general)

Updated the title, per your suggestion. Thanks!

lianetm commented:
Cherry-picked to 4.2 https://github.com/apache/kafka/commit/9ced615194aae5998a65a9626afdff04dc3760b9

chia7712 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java:
```diff
@@ -879,8 +879,26 @@ public synchronized Set<TopicPartition> partitionsNeedingReset(long nowMs) {
+        return result;
+    }
+
+    public synchronized boolean hasPartitionsNeedingValidation(long nowMs) {
```
not sure why this method is not used.

kirktrue replied:
I will submit a PR to remove it.

chia7712 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PositionsValidator.java:
```diff
@@ -0,0 +1,151 @@
+     * If we have seen new metadata (as tracked by {@link org.apache.kafka.clients.Metadata#updateVersion()}), then
+     * we should check that all the assignments have a valid position.
+     */
+    void validatePositionsOnMetadataChange(ApiVersions apiVersions) {
```
Excuse me, why this method `validatePositionsOnMetadataChange` does not need to call `maybeThrowError`?

chia7712 reviewed (commented):
@kirktrue Thanks you for the optimization. Looking at https://github.com/apache/kafka/pull/20324#issuecomment-3172338727, is it true that `bytes-consumed-rate` metrics still indicates a clear performance regression, even with this patch?

kirktrue commented:
> Looking at [#20324 (comment)](https://github.com/apache/kafka/pull/20324#issuecomment-3172338727), is it true that `bytes-consumed-rate` metrics still indicates a clear performance regression, even with this patch?

Unfortunately, that benchmark is quite outdated and misleading. We're using a different set of benchmarks using an internal tool that shows the gap as being much closer between the `CLASSIC` and `CONSUMER` group protocols.

chia7712 commented:
>  We're using a different set of benchmarks using an internal tool that shows the gap as being much closer between the CLASSIC and CONSUMER group protocols.

That is very unfortunate. I won't be able to access this insightful data analysis :smile:
